timemachine/cpp/src/nonbonded_interaction_group.hpp

```diff
@@ -0,0 +1,125 @@
+#pragma once
+
+#include "neighborlist.hpp"
+#include "potential.hpp"
+#include "vendored/jitify.hpp"
+#include <array>
+#include <set>
+#include <vector>
+
+namespace timemachine {
+
+typedef void (*k_nonbonded_fn)(
+    const int NC,
+    const int NR,
+    const double *__restrict__ coords,
+    const double *__restrict__ params, // [N]
+    const double *__restrict__ box,
+    const double *__restrict__ dl_dp,
+    const double *__restrict__ coords_w, // 4D coords
+    const double *__restrict__ dw_dl,    // 4D derivatives
+    const double beta,
+    const double cutoff,
+    const int *__restrict__ ixn_tiles,
+    const unsigned int *__restrict__ ixn_atoms,
+    unsigned long long *__restrict__ du_dx,
+    unsigned long long *__restrict__ du_dp,
+    unsigned long long *__restrict__ du_dl_buffer,
+    unsigned long long *__restrict__ u_buffer);
+
+template <typename RealType, bool Interpolated> class NonbondedInteractionGroup : public Potential {
+
+private:
+    const int N_;  // N_ = NC_ + NR_
+    const int NR_; // number of row atoms
+    const int NC_; // number of column atoms
+
+    std::array<k_nonbonded_fn, 16> kernel_ptrs_;
+
+    unsigned int *d_col_atom_idxs_;
+    unsigned int *d_row_atom_idxs_;
+
+    int *d_lambda_plane_idxs_;
+    int *d_lambda_offset_idxs_;
+    int *p_ixn_count_; // pinned memory
+
+    double beta_;
+    double cutoff_;
+    Neighborlist<RealType> nblist_;
+
+    double nblist_padding_;
+    double *d_nblist_x_;    // coords which were used to compute the nblist
+    double *d_nblist_box_;  // box which was used to rebuild the nblist
+    int *d_rebuild_nblist_; // whether or not we have to rebuild the nblist
+    int *p_rebuild_nblist_; // pinned
+
+    unsigned int *d_perm_; // hilbert curve permutation
+
+    double *d_w_;     //
+    double *d_dw_dl_; //
+
+    double *d_sorted_x_;     //
```

Comment:
Are these empty comments suppose to be filled out?

Reply:
These are carried over from `nonbonded_all_pairs.cu`. We should probably clean these up at some point; I was thinking to do some of this in a separate PR that also DRYs some of the common bits.

Reply:
Added to #639